test/model/message_test.dart

```diff
@@ -0,0 +1,418 @@
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/message_list.dart';
+import 'package:zulip/model/narrow.dart';
+import 'package:zulip/model/store.dart';
+
+import '../api/fake_api.dart';
+import '../api/model/model_checks.dart';
+import '../example_data.dart' as eg;
+import '../stdlib_checks.dart';
+import 'message_list_test.dart';
+import 'store_checks.dart';
+import 'test_store.dart';
+
+void main() {
+  // These "late" variables are the common state operated on by each test.
+  // Each test case calls [prepare] to initialize them.
+  late Subscription subscription;
+  late PerAccountStore store;
+  late FakeApiConnection connection;
+  // [messageList] is here only for the sake of checking when it notifies.
+  // For anything deeper than that, use `message_list_test.dart`.
+  late MessageListView messageList;
+  late int notifiedCount;
+
+  void checkNotified({required int count}) {
+    check(notifiedCount).equals(count);
+    notifiedCount = 0;
+  }
+  void checkNotNotified() => checkNotified(count: 0);
+  void checkNotifiedOnce() => checkNotified(count: 1);
+
+  /// Initialize [store] and the rest of the test state.
+  Future<void> prepare({Narrow narrow = const CombinedFeedNarrow()}) async {
+    final stream = eg.stream();
+    subscription = eg.subscription(stream);
+    store = eg.store();
+    await store.addStream(stream);
+    await store.addSubscription(subscription);
+    connection = store.connection as FakeApiConnection;
+    notifiedCount = 0;
+    messageList = MessageListView.init(store: store, narrow: narrow)
+      ..addListener(() {
+        notifiedCount++;
+      });
+    check(messageList).fetched.isFalse();
+    checkNotNotified();
+  }
+
+  /// Perform the initial message fetch for [messageList].
+  ///
+  /// The test case must have already called [prepare] to initialize the state.
+  Future<void> prepareMessages(
+    List<Message> messages, {
+    bool foundOldest = false,
+  }) async {
+    connection.prepare(json:
+      newestResult(foundOldest: foundOldest, messages: messages).toJson());
+    await messageList.fetchInitial();
+    checkNotifiedOnce();
+  }
+
+  Future<void> addMessages(Iterable<Message> messages) async {
+    for (final m in messages) {
+      await store.handleEvent(MessageEvent(id: 0, message: m));
+    }
+    checkNotified(count: messageList.fetched ? messages.length : 0);
+  }
+
+  group('reconcileMessages', () {
+    test('from empty', () async {
+      await prepare();
+      check(store.messages).isEmpty();
+      final message1 = eg.streamMessage();
+      final message2 = eg.streamMessage();
+      final message3 = eg.dmMessage(from: eg.otherUser, to: [eg.selfUser]);
+      final messages = [message1, message2, message3];
+      store.reconcileMessages(messages);
+      check(messages)
+        .deepEquals([message1, message2, message3].map((m) => (it) => identical(it, m)));
```

Comment:
Hmm, these callbacks don't look right. The `deepEquals` doc says (via a macro/template from `deepCollectionEquals`:
```dart
/// Elements, keys, or values within [expected] which are `Condition` callbacks
/// are run against the value in the same position within [actual].
/// Condition callbacks must take a `Subject<Object?>` or `Subject<dynamic>` and
/// may not use a more specific generic.
/// Use `(Subject<Object?> s) => s.isA<Type>()` to check expectations for
/// specific element types.
/// Note also that the argument type `Subject<Object?>` cannot be inferred and
/// must be explicit in the function definition.
```

So the `it` argument should be a Subject, not a Message itself.

In fact, the test still passes if I write:
```diff
       check(messages)
-        .deepEquals([message1, message2, message3].map((m) => (it) => identical(it, m)));
+        .deepEquals([message1, message2, message3].map((m) => (it) => !identical(it, m)));
```

Definitely something to fix there, then :slightly_smiling_face:

Reply:
OK, and after #F719 I think the analyzer should complain on this line.